edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/MetricsWorker.cs

```diff
@@ -53,38 +57,67 @@ internal async Task Scrape(CancellationToken cancellationToken)
         }
 
         internal async Task Upload(CancellationToken cancellationToken)
+        {
+            if (!await this.TryUploadAndClear(cancellationToken))
+            {
+                await this.BeginUploadRetry(cancellationToken);
+            }
+        }
+
+        async Task<bool> TryUploadAndClear(CancellationToken cancellationToken)
         {
             using (await this.scrapeUploadLock.LockAsync(cancellationToken))
             {
                 Log.LogInformation($"Uploading Metrics");
-                IEnumerable<Metric> metricsToUpload = await this.storage.GetAllMetricsAsync();
-                metricsToUpload = RemoveDuplicateMetrics(metricsToUpload);
-                await this.uploader.PublishAsync(metricsToUpload, cancellationToken);
-                Log.LogInformation($"Uploaded Metrics");
-                await this.storage.RemoveAllReturnedMetricsAsync();
+                IEnumerable<Metric> metricsToUpload = (await this.storage.GetAllMetricsAsync()).CondenseTimeSeries();
+
+                bool uploaded;
+                try
+                {
+                    uploaded = await this.uploader.PublishAsync(metricsToUpload, cancellationToken);
+                }
+                catch (Exception ex)
+                {
+                    // If unexpected error publishing metrics, delete stored ones to prevent storage buildup.
+                    Log.LogError($"Unexpected error publishing metrics. Deleting stored metrics.");
+                    await this.storage.RemoveAllReturnedMetricsAsync();
+                    throw ex;
+                }
+
+                if (uploaded)
```

Comment:
Why not just move all of this into the try? Then you wouldn't need the bool uploaded declared out on line 74? Doesn't seem like it buys you anything having this code out here.

Reply:
This is for lines 87-97